File: Matlab/sysdyn.m
function [x_p,sysvar] = sysdyn(x,t,dt,iter,syspar,sysvar)
% x - current system state
% t - current instant of time
% dt - time step
% iter - iteration during time-step integration (e.g. iter = 1, 2, 3, 4 for 4th-order Runge-Kutta)
% syspar - structure with system parameters

%% Read state vector
eta = x(1:2);
nu = x(3:4);

%% Equations of motions
Mrb = syspar.strprop.Mrb;
Ainf = syspar.hydro.Ainf;
Dl = syspar.hydro.Dl;
Dq = syspar.hydro.Dq;
Kret = syspar.hydro.Kret;
Chs = syspar.hydro.Chs;
Cmr = syspar.external.Cmr;

% Evaluate Cummins integral
ktime = sysvar.k;
if iter == 1    
    thist = sysvar.t(1:ktime);
    nuhist = sysvar.nu(:,1:ktime);
    [mu] = cumminsint(Kret,thist,nuhist);
    sysvar.mu(:,ktime) = mu;
else
    mu = sysvar.mu(:,ktime);
end
% mu = [0;0];

% External loads
Fext = [0;0]; % Not implemented - keep zero for now

% System dynamics
eta_p = nu;
nu_p = (Mrb+Ainf)\(-Dl*nu-Dq*abs(nu).*nu-mu-(Chs+Cmr)*eta+Fext);
% nu_p(1) = (Mrb(1,1)+Ainf(1,1))\(-Dl(1,1)*nu(1)-mu(1)-(Chs(1,1)+Cmr(1,1))*eta(1)+Fext(1));
% nu_p(2) = 0;


%% Define y_p
x_p = [eta_p;nu_p];

File: Fortran/sysdyn.f90
FUNCTION sysdyn(x,t,dt,iter)

IMPLICIT NONE

INTEGER                                   :: iter        ! Current RK iteration
INTEGER                                   :: N           ! Order of system

REAL                                      :: t           ! Time instant
REAL                                      :: dt          ! Time step
REAL, DIMENSION(2)                        :: x           ! Current state
REAL, DIMENSION(2)                        :: sysdyn      ! Time derivative of x
REAL                                      :: eta         ! Position state
REAL                                      :: nu          ! Velocity state
!REAL, DIMENSION(:), ALLOCATABLE           :: eta         ! Position state
!REAL, DIMENSION(:), ALLOCATABLE           :: nu          ! Velocity state
REAL                                      :: eta_p       ! Time derivative of position state
REAL                                      :: nu_p        ! Time derivative of velocity state
!REAL, DIMENSION(:), ALLOCATABLE           :: eta_p       ! Time derivative of position state
!REAL, DIMENSION(:), ALLOCATABLE           :: nu_p        ! Time derivative of velocity state
REAL                                      :: Mrb         ! Rigid-body inertia matrix
REAL                                      :: A0          ! Zero-frequency added mass matrix
REAL                                      :: Dl          ! Linear viscous damping matrix
REAL                                      :: Cmr         ! Mooring stiffness matrix
!REAL, DIMENSION(1,1)                      :: Mrb         ! Rigid-body inertia matrix
!REAL, DIMENSION(1,1)                      :: A0          ! Zero-frequency added mass matrix
!REAL, DIMENSION(1,1)                      :: Dl          ! Linear viscous damping matrix
!REAL, DIMENSION(1,1)                      :: Cmr         ! Mooring stiffness matrix

N = 2

Mrb = 7794048
A0 = 7.99e6
Dl = 1e5
Cmr = 41180

eta = x(1)
nu = x(2)

eta_p = nu
nu_p = (-Dl*nu-Cmr*eta)/(Mrb+A0)

sysdyn(1) = eta_p
sysdyn(N/2+1:N) = nu_p

END FUNCTION sysdyn